lib/datadog/tracing/contrib/grpc/formatting.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Tracing
+    module Contrib
+      module GRPC
+        module Formatting
+          VALUE_UNKNOWN = 'unknown'
+
+          # A class to extract GRPC span attributes from the GRPC implementing class method object.
+          class MethodObjectFormatter
+            # grpc_full_method is a string containing all the rpc method information (from the Protobuf definition)
+            # in a single string with the following format: /$package.$service/$method
+            attr_reader :grpc_full_method
+
+            # legacy_grpc_service is built using the Ruby GRPC service implementation package and class name instead
+            # of the rpc interface representation from Protobuf. It's kept for compatibility.
+            attr_reader :legacy_grpc_service
+
+            # legacy_grpc_method is built using the Ruby GRPC service implementation method name instead of the rpc
+            # interface representation from Protobuf. It's kept for compatibility.
+            attr_reader :legacy_grpc_method
+
+            # resource_name is used for the span resource name.
+            attr_reader :resource_name
+
+            def initialize(grpc_method_object)
+              @grpc_full_method = format_full_method(grpc_method_object)
+              @resource_name = format_resource_name(grpc_method_object)
+              @legacy_grpc_method = extract_legacy_grpc_method(grpc_method_object)
+              @legacy_grpc_service = extract_legacy_grpc_service(grpc_method_object)
+            end
+
+            private
+
+            def format_full_method(grpc_method_object)
+              service = extract_grpc_service(grpc_method_object)
+              method = extract_grpc_method(grpc_method_object)
+              "/#{service}/#{method}"
+            end
+
+            def extract_grpc_service(grpc_method_object)
+              owner = grpc_method_object.owner
+              return VALUE_UNKNOWN unless owner.instance_variable_defined?(:@service_name)
+
+              # Ruby protoc generated code includes this variable which directly contains the value from the original
+              # protobuf definition
+              owner.service_name.to_s
+            end
+
+            # extract_grpc_method attempts to find the original method name from the Protobuf file definition,
+            # since grpc gem forces the implementation method name to be in snake_case.
+            def extract_grpc_method(grpc_method_object)
+              owner = grpc_method_object.owner
+
+              return VALUE_UNKNOWN unless owner.instance_variable_defined?(:@rpc_descs)
```

Comment:
I added this check because for the case of the previous grpc Ruby class (manual) implementation this was not defined, and this is added by the protoc Ruby generated one.

Even though I just checked the [grpc gem](https://github.com/grpc/grpc/blob/2d4f3c56001cd1e1f85734b2f7c5ce5f2797c38a/src/ruby/lib/grpc/generic/rpc_server.rb#L351-L360) fails to start if you provide a class without this or this having no items.

Would you prefer to remove this check? (in that case I think we probably need to update some unit tests that are relying on a non-autogenerated grpc test service class if I recall correctly)